### PR TITLE
NodeScopeResolver: do not corrupt type info on invalid offset assignments

### PR DESCRIPTION
**Pull request by dktapps**
this fixes phpstan/phpstan#3782.
I'm not sure if this may have any side effects, but I can't see any circumstances where it makes sense to reassign an object as an array on array dim fetch.

### CODE REVIEW

**Comment by dktapps** [comment]
The latest commit caused a change in behaviour to the result of string offset accesses, which I believe is acceptable because the old behaviour was causing some errors not to be reported, like the one missing from line 8 here: https://phpstan.org/r/801fa88f-87bb-49c3-8c6c-ea8b781fb950

File: tests/PHPStan/Analyser/data/bug-3782.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+			assertType('1', $var);
+		}
+	}
+
+	public static function sayHello4(array $data): void
+	{
+		$var = "hello world";
+		foreach($data as $key => $value){
+			$var[$key] = $value;
+			assertType('\'hello world\'', $var);
```
Comment (dktapps) [inline tests/PHPStan/Analyser/data/bug-3782.php:41]: This should actually be `string` (see phpstan/phpstan#3803). Not sure if that should be addressed here or not.

**Comment by dktapps** [comment]
This is fixed for the object case, but the following array coercion is still not expected (and is fixed by this PR): https://phpstan.org/r/43b7a9c7-06cf-414c-ab58-63392fa4a357
https://3v4l.org/N9Aoe

**Comment by ondrejmirtes** [comment]
This happens because `$str` after the first assignment becomes ErrorType.

**Comment by ondrejmirtes** [comment]
The linked issue got fixed.

**Pull request closed without merging**